Tabix support for IndexFeatureFile, and move to htsjdk 2.6.1

magicDGS opened the pull request:
* Update htsjdk to 2.6.1
* Fix #2080

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java:
```diff
@@ -104,13 +105,20 @@ private File determineFileName(final Index index) {
+                return IndexFactory
+                        .createIndex(featureFile, codec, IndexFactory.IndexType.TABIX, null);
+            } catch(TribbleException.MalformedFeatureFile e) {
+                throw e;
```
Wrap this in a `UserException.CouldNotReadInputFile`

magicDGS replied:
This changed the behaviour and that's why I used the double catch clause. Should I unify the catch?

magicDGS replied:
Done, but wrap into `UserException.MalformedFile` instead. Changed also in the tests.

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java:
```diff
@@ -125,6 +133,11 @@ else if ( featureFile.getName().endsWith(GVCF_FILE_EXTENSION) ) {
+     * @deprecated because it is not used anymore
+     */
+    // TODO: should be removed?
+    @Deprecated
```
It's private, so go ahead and delete it if it's unused.

magicDGS replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/IndexFeatureFile.java:
```diff
@@ -104,13 +105,20 @@ private File determineFileName(final Index index) {
+                // TODO: this could benefit from provided sequence dictionary from reference
+                // TODO: this can be an optional parameter for the tool
+                return IndexFactory
+                        .createIndex(featureFile, codec, IndexFactory.IndexType.TABIX, null);
```
Add integration tests showing that we can now index block-compressed feature files that are not vcfs (eg., BED files, SAMPileup files).

magicDGS replied:
Do you have any suggestion for the TODO? I think that an optional parameter with the sequence dictionary will be interesting for fast indexing...

magicDGS replied:
Added the tests, but the one for bed files is not enabled.

magicDGS commented:
I addressed the comments, @droazen. But I found that the way to detect the codec for the files is relying on `canDecode`. But even if the a block-compressed file could be read by a codec through the `AbstractFeatureReader`, the `canDecode` for bed files returns false for block-compressed extension.

This should be change at the htsjdk level if GATK is planing to read block-compressed feature inputs. The same for the codecs implemented in the framework (I did it for the `SAMPileupCodec`), but for instance the table codec could not be used in a compressed way because you can't create a tabix index for it...

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/ProgressReportingDelegatingCodec.java:
```diff
@@ -122,4 +123,12 @@ public boolean canDecode(final String path) {
+    @Override
+    public TabixFormat getTabixFormat() {
+        if (delegatee == null) {
+            throw new IllegalArgumentException("this codec cannot be used without a delegatee.");
```
Should be `IllegalStateException` (since there are no arguments here).

magicDGS replied:
Sure!

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
     }
 
+    // TODO: this is not enabled because canDecode returns false for bed.gz extension
+    // TODO: should this be changed at the htsjdk level to return true also for a block-compressed extension?
```
Yes, you're right -- assuming that `AbstractFeatureReader.getFeatureReader()` will work correctly in the block-gzipped case with this codec, `canDecode()` should be returning true. Recommend filing a ticket or PR against htsjdk to fix this.

magicDGS replied:
Done (https://github.com/samtools/htsjdk/pull/704)

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
+    // TODO: should this be changed at the htsjdk level to return true also for a block-compressed extension?
+    @Test(enabled = false)
+    public void testBedGZIndex() {
+        final File ORIG_FILE = getTestFile("test_bed_for_index.bed.gz");
```
Is this test file gzipped or block-gzipped? (they are not the same thing!)

magicDGS replied:
I bgzip this and the pileup.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
+    @Test(enabled = false)
+    public void testBedGZIndex() {
+        final File ORIG_FILE = getTestFile("test_bed_for_index.bed.gz");
+        final File outName = BaseTest.createTempFile("test_bed_for_index.bed.", TabixUtils.STANDARD_INDEX_EXTENSION);
```
Can just write `createTempFile()` instead of `BaseTest.createTempFile()`, since it's inherited.

magicDGS replied:
Done. I change it also in the rest of the tests in this file.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
+        Assert.assertEquals(res, outName.getAbsolutePath());
+
+        final Index index = IndexFactory.loadIndex(res.toString());
+        Assert.assertTrue(index instanceof LinearIndex);
```
This assert doesn't seem right -- it should be a `TabixIndex`, shouldn't it?

magicDGS replied:
Sorry, copy-paste error. Fixed while extracting the common code from both tests.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
+
+    @Test
+    public void testSAMPileupGZIndex() {
+        final File ORIG_FILE = getTestFile("test_sampileup_for_index.pileup.gz");
```
Same question here -- is this file block-gzipped, or regular gzipped?

magicDGS replied:
I bgzip this and the bed file. If it is not block-compressed, the test will fail (like in `testVCFGZIPIndex()` and `testVCFGZIPIndex_inferredName()`. I added a comment to point it out.

droazen commented on src/test/java/org/broadinstitute/hellbender/tools/IndexFeatureFileIntegrationTest.java:
```diff
@@ -265,6 +265,61 @@ public void testBedIndex() {
+            Assert.assertFalse(index.containsChromosome(chr), String.valueOf(chr));
+        }
+
+        Assert.assertEquals(index.getSequenceNames(), Arrays.asList("1", "2", "3", "4"));
```
The test methods here have a lot of repeated code between them -- perhaps you could extract a shared private method that takes the expected index class as an argument?

magicDGS replied:
Sure, I extracted this and the pileup code.

magicDGS replied:
The case of the testSAMPileupGZIndex is different, because the expected chromosomes are not the same (3 is included)...

droazen commented:
Second-pass review complete -- back to @magicDGS

magicDGS commented:
Adressed the comments -- back to you @droazen. I will squash/rebase when you tell me.

droazen commented:
:+1: looks good, I've merged this (github's new "squash and merge" feature eliminates the need for PR authors to manually squash at the end of review in most cases)

magicDGS commented:
Thanks @droazen!

vdauwera commented:
@droazen I know it's too late now but for the future... I would still recommend requiring the author to do a manual rebase to clean up the commit history, otherwise later it's going to be a pain to do things like assemble coherent and succinct version notes. We end up with many lines in the logs that consist of "addressing comments" and "fixing tests", which are completely uninformative.

vdauwera commented:
Alright, @droazen just showed me what actually got committed and it's all cleaned up. Yay! So now I approve this feature. As long as people do take advantage of the opportunity to clean up the squashed commit message!
